[Java] Support GraalVM Native Image

## Is your feature request related to a problem? Please describe.

Java works but compile to native java doesn't work.

Native Java:
https://www.graalvm.org/

## Describe the solution you'd like
Add native java support by adding classes that need reflection to meta files.

## Additional context

Exception when run in native java.

java.lang.RuntimeException: java.lang.UnsupportedOperationException: java.lang.NoSuchMethodException: java.util.concurrent.ConcurrentSkipListSet.<init>(java.util.Comparator)
	at com.aigens.api.TestApi.hello(TestApi.java:85)
	at com.aigens.api.TestApi$quarkusrestinvoker$hello_a988a40893ff99606bf40d6c00ea871031f3749c.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.5/java.lang.Thread.run(Thread.java:833)
	at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:775)
	at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:203)
Caused by: java.lang.UnsupportedOperationException: java.lang.NoSuchMethodException: java.util.concurrent.ConcurrentSkipListSet.<init>(java.util.Comparator)
	at io.fury.serializer.CollectionSerializers$SortedSetSerializer.<init>(CollectionSerializers.java:521)
	at io.fury.serializer.CollectionSerializers$ConcurrentSkipListSetSerializer.<init>(CollectionSerializers.java:707)
	at io.fury.serializer.CollectionSerializers.registerDefaultSerializers(CollectionSerializers.java:918)
	at io.fury.resolver.ClassResolver.addDefaultSerializers(ClassResolver.java:309)
	at io.fury.resolver.ClassResolver.initialize(ClassResolver.java:297)
	at io.fury.Fury.<init>(Fury.java:128)

--- a/README.md
+++ b/README.md
@@ -26,10 +26,11 @@ In addition to cross-language serialization, Fury also features at:
 
 - Drop-in replace Java serialization frameworks such as JDK/Kryo/Hessian without modifying any code, but 100x faster. 
   It can greatly improve the efficiency of high-performance RPC calls, data transfer, and object persistence.
-- **100% compatible** with JDK serialization: supporting JDK `writeObject/readObject/writeReplace/readResolve/readObjectNoData/Externalizable` serialization API. 
+- **100% compatible** with JDK serialization API with much faster implementation: supporting JDK `writeObject/readObject/writeReplace/readResolve/readObjectNoData/Externalizable` API. 
 - Supports **Java 8~21**, Java 17+ `record` is supported too.
+- Support AOT compilation serialization for [GraalVM native image](docs/guide/graalvm_guide.md), and no reflection/serialization json config are needed.
 - Supports shared and circular reference object serialization for golang.
-- Supports [scala serialization](https://github.com/alipay/fury/blob/main/docs/guide/scala_guide.md)
+- Supports [scala serialization](docs/guide/scala_guide.md)
 - Supports automatic object serialization for golang.
 
 ## Protocols

--- a/docs/README.md
+++ b/docs/README.md
@@ -1,9 +1,10 @@
 # User Guide
-- For Cross Language Object Graph Guide, see [xlang_object_graph_guide](https://github.com/alipay/fury/blob/main/docs/guide/xlang_object_graph_guide.md) doc.
-- For Java Object Graph Guide, see [java_object_graph_guide](https://github.com/alipay/fury/blob/main/docs/guide/java_object_graph_guide.md) doc.
-- For Row Format Guide, see [row format_guide](https://github.com/alipay/fury/blob/main/docs/guide/row_format_guide.md) doc.
+- For Cross Language Object Graph Guide, see [xlang_object_graph_guide](guide/xlang_object_graph_guide.md) doc.
+- For Java Object Graph Guide, see [java_object_graph_guide](guide/java_object_graph_guide.md) doc.
+- For Row Format Guide, see [row format_guide](guide/row_format_guide.md) doc.
 - For Scala Guide, see [scala_guide](guide/scala_guide.md) doc.
+- For using Fury with GraalVM native image, see [graalvm native image guide](guide/graalvm_guide.md) doc.
 
 # Development
-- For cpp debug, see [cpp_debug](https://github.com/alipay/fury/blob/main/docs/cpp_debug.md) doc.
-- For development, see [CONTRIBUTING](https://github.com/alipay/fury/blob/main/CONTRIBUTING.md) doc.
+- For cpp debug, see [cpp_debug](cpp_debug.md) doc.
+- For development, see [CONTRIBUTING](../CONTRIBUTING.md) doc.

--- a/docs/guide/DEVELOPMENT.md
+++ b/docs/guide/DEVELOPMENT.md
@@ -1,6 +1,6 @@
 <!-- fury_frontmatter --
 title: Development
-order: 5
+order: 6
 -- fury_frontmatter -->
 # How to build to Fury
 

--- a/docs/guide/graalvm_guide.md
+++ b/docs/guide/graalvm_guide.md
@@ -1,0 +1,116 @@
+<!-- fury_frontmatter --
+title: GraalVM Guide
+order: 6
+-- fury_frontmatter -->
+
+# GraalVM Native Image
+GraalVM native image can compile java code into native code ahead into a binary for building faster, smaller, leaner applications.
+Fury run on GraalVM native image pretty well.
+Although GraalVM native image doesn't have a JIT compiler to compile bytecode into machine code, and doesn't support 
+reflection unless configure reflection file. Fury can generate all serializer code for fury JIT framework and MethodHandler/LambdaMetafactory at graalvm build time. Then use those generate code for serialization at runtime without 
+any extra cost, the performance will be pretty goog..
+
+In order to use Fury on graalvm native image, you must create Fury as an static field of a class, and register all classes at
+ the enclosing class initialize time. Then configure `native-image.properties` under 
+`resources/META-INF/native-image/$xxx/native-image.propertie` to tell graalvm to init the enclosing class at native image 
+build time. For example, here we configure `io.fury.graalvm.Example` class be init at build time:
+```properties
+Args = --initialize-at-build-time=io.fury.graalvm.Example
+```
+
+Another benefit using fury is that you don't have to configure [reflection json](https://www.graalvm.org/latest/reference-manual/native-image/metadata/#specifying-reflection-metadata-in-json) and 
+[serialization json](https://www.graalvm.org/latest/reference-manual/native-image/metadata/#serialization), which is
+very tedious, cumbersome and inconvenient. WHen using fury, you just need to invoke 
+`io.fury.Fury.register(Class<?>, boolean)` for every type you want to serialize.
+
+## Not thread safe fury
+Example:
+```java
+import io.fury.Fury;
+import io.fury.util.Preconditions;
+
+import java.util.List;
+import java.util.Map;
+
+public class Example {
+  public record Record (
+    int f1,
+    String f2,
+    List<String> f3,
+    Map<String, Long> f4) {
+  }
+
+  static Fury fury;
+
+  static {
+    fury = Fury.builder().build();
+    // register and generate serializer code.
+    fury.register(Record.class, true);
+  }
+
+  public static void main(String[] args) {
+    Record record = new Record(10, "abc", List.of("str1", "str2"), Map.of("k1", 10L, "k2", 20L));
+    System.out.println(record);
+    byte[] bytes = fury.serialize(record);
+    Object o = fury.deserialize(bytes);
+    System.out.println(o);
+    Preconditions.checkArgument(record.equals(o));
+  }
+}
+```
+Then add `io.fury.graalvm.Example` build time init to `native-image.properties` configuration:
+```properties
+Args = --initialize-at-build-time=io.fury.graalvm.Example
+```
+
+## Thread safe fury
+```java
+import io.fury.Fury;
+import io.fury.ThreadLocalFury;
+import io.fury.ThreadSafeFury;
+import io.fury.util.Preconditions;
+
+import java.util.List;
+import java.util.Map;
+
+public class ThreadSafeExample {
+  public record Foo (
+    int f1,
+    String f2,
+    List<String> f3,
+    Map<String, Long> f4) {
+  }
+  
+  static ThreadSafeFury fury;
+
+  static {
+    fury = new ThreadLocalFury(classLoader -> {
+      Fury f = Fury.builder().build();
+      // register and generate serializer code.
+      f.register(Foo.class, true);
+      return f;
+    });
+  }
+
+  public static void main(String[] args) {
+    System.out.println(fury.deserialize(fury.serialize("abc")));
+    System.out.println(fury.deserialize(fury.serialize(List.of(1,2,3))));
+    System.out.println(fury.deserialize(fury.serialize(Map.of("k1", 1, "k2", 2))));
+    Foo foo = new Foo(10, "abc", List.of("str1", "str2"), Map.of("k1", 10L, "k2", 20L));
+    System.out.println(foo);
+    byte[] bytes = fury.serialize(foo);
+    Object o = fury.deserialize(bytes);
+    System.out.println(o);
+  }
+}
+```
+Then add `io.fury.graalvm.ThreadSafeExample` build time init to `native-image.properties` configuration:
+```properties
+Args = --initialize-at-build-time=io.fury.graalvm.ThreadSafeExample
+```
+
+## Framework Integration
+For framework developers, if you want to integrate fury for serialization, you can provided a configuration file to let 
+the users to list all the classes they want to serialize, then you can load those classes and invoke 
+`io.fury.Fury.register(Class<?>, boolean)` to register those classes in your Fury integration class, and configure that 
+class be initialized at graalvm native image build time.

